html: Enhance media network and decoding error handling

tharkum opened the pull request:
The 'media' element has multiple media data processing steps for media source loading and requires properly handling any media network and decoding errors between different components:
- element (resource selection algorithm)
- fetch context/listener (async response)
- media backend (async playback events)

The fetch listener will stop processing any actions (response/chunk/eof) in case if current fetch request was cancelled (due to a network/decoding error or it was aborted by user).

Also send EOS event to media backend ONLY once the entire media resource has been fetched so media pipeline will not generate decoding (playback) error for fetched empty sources (eg. "about:blank" or "data:,").

See https://html.spec.whatwg.org/multipage/#media-data-processing-steps-list

Testing: No changes in WPT test expectations

Fixes: https://github.com/servo/servo/issues/39759 ("about:blank")
Fixes: https://github.com/servo/servo/issues/38980 ("data:,")

tharkum committed:
html: Enhance media network and decoding error handling

The 'media' element has multiple media data processing steps
for media source loading and requires properly handling any media
network and decoding errors between different components:
- element (resource selection algorithm)
- fetch context/listener (async response)
- media backend (async playback events)

The fetch listener will stop processing any actions
(response/chunk/eof) in case if current fetch request was cancelled
(due to a network/decoding error or it was aborted by user).

Also send EOS event to media backend ONLY once the entire media
resource has been fetched so media pipeline will not generate
decoding (playback) error for fetched empty sources
(eg. "about:blank" or "data:,").

See https://html.spec.whatwg.org/multipage/#media-data-processing-steps-list

Testing: No changes in WPT test expectations

Fixes: https://github.com/servo/servo/issues/39759 ("about:blank")
Fixes: https://github.com/servo/servo/issues/38980 ("data:,")

Signed-off-by: Andrei Volykhin <andrei.volykhin@gmail.com>

jdm commented on components/script/dom/html/htmlmediaelement.rs:
```diff
@@ -1637,6 +1637,54 @@ impl HTMLMediaElement {
+        self.error.set(Some(&*MediaError::new(
+            &self.owner_window(),
+            error,
+            CanGc::note(),
```
Please add a `CanGc` argument to this method and use that instead of each use of CanGc::note() in this method.

jdm reviewed (changes requested):
Looks good, but I'd like to take another look after the CanGc-related changes have been made!

tharkum commented on components/script/dom/html/htmlmediaelement.rs:
```diff
@@ -2250,26 +2249,22 @@ impl HTMLMediaElement {
+        // The player needs more data. If we already have a valid fetch request, we do nothing.
+        // Otherwise, if we have no request and the previous request was cancelled because we got an
+        // EnoughData event, we restart fetching where we left.
+        if let Some(ref mut current_fetch_context) = *self.current_fetch_context.borrow_mut() {
```
Still under discussion, asked the question in DM (Zulip)

tharkum commented:
- Added `#[no_trace]` attribute to `current_fetch_context` member
  `HTMLMediaElementFetchContext` have no DOM traceable members (except `FetchCanceller`, but it is also have no DOM traceable members), so mutable borrowing within GC should be safe (`self.current_fetch_context.borrow_mut() + seek()`)
```
pub(crate) struct FetchCanceller {
    #[no_trace]
    request_id: Option<RequestId>,
    #[no_trace]
    core_resource_thread: Option<CoreResourceThread>,
}
```
- Replaced `DomRefCell` to `RefCell` for `HTMLMediaElementFetchContext` and `BufferedDataSource`
  None of specific functions for `DomRefCell` were used for these members

@jdm < Please take a look. The fine-grained borrowing still could be added as well...

tharkum commented:
- Rebased on top of main branch (incl. addition of `CanGc` to methods)

jdm commented:
I'll have a closer look tonight, but I'm worried that fixes like marking fields `#[no_trace]` to allow mutable borrowing during a GC both make our `CanGC` scheme harder to rely upon, and pose a risk if we introduce a transitive GC member to the field in the future.

tharkum commented:
Of course there are some risk to implement it wrong with introduced a transitive GC member to the field in future, but it is trade off between performance and safety (no point to trace non js manageable objects during CG).

BTW, I think `FetchCanceller` shouldn't be `JSTraceable` which will reduce path of GC tracing for some objects (like `InProgressLoad`, `DocumentLoader`, `DroppableEventSource`)

```
#[derive(Default, JSTraceable, MallocSizeOf)]
pub(crate) struct FetchCanceller {
    #[no_trace]
    request_id: Option<RequestId>,
    #[no_trace]
    core_resource_thread: Option<CoreResourceThread>,
}
```

jdm commented on components/script/dom/html/htmlmediaelement.rs:
```diff
@@ -2252,26 +2249,22 @@ impl HTMLMediaElement {
+        // Otherwise, if we have no request and the previous request was cancelled because we got an
+        // EnoughData event, we restart fetching where we left.
+        if let Some(ref mut current_fetch_context) = *self.current_fetch_context.borrow_mut() {
+            if let Some(cancel_reason) = current_fetch_context.cancel_reason() {
```
As far as I can see, the changes to this part of the code are not required to solve the original issue, right? If that's true, I would prefer to revert them as well as the changes to make `current_fetch_context` use `#[no_trace]`, because I'm still not confident that that's a good long-term solution.
